lua/wire/stools/expression2.lua

```diff
@@ -55,20 +72,135 @@ if SERVER then
-			self:Download(player, trace.Entity)
-			return true
+	local function RequestView(chip, initiator)
+		local index = chip:EntIndex()
+		local truncName = string.sub(chip.name, 1, 256) -- In case someone starts making cursed names
+
+		-- Make sure this isn't creating a request for a chip with an outstanding valid request
+		if viewRequests[index] and viewRequests[index].expiry > CurTime() then
+			BetterChatPrint(initiator, "Request to view '"..truncName.."' already sent")
+			return
 		end
 
-		net.Start("WireExpression2_OpenEditor") net.Send(player)
-		return false
+		-- Otherwise, print to the tool user's chat that a view request was sent
+		-- and send a view request to the chip's owner (also print a message to their chat to tell them they received a request)
+		BetterChatPrint(initiator, "E2 view request sent for '"..truncName.."' owned by "..chip.player:Nick())
+		BetterChatPrint(chip.player, "You just received a request to view your E2 '"..truncName.."' from "..initiator:Nick()..", which you can view in your context menu")
+
+		-- Add the request data to the local requests table for when the request from the client comes in
+		viewRequests[index] = {
+			initiator = initiator,
+			owner = chip.player,
+			name = truncName,
+			chip = chip,
+			expiry = CurTime() + 60 -- 1 minute for the request before it's invalidated (could make this a convar)
+		}
+
+		net.Start("WireExpression2_ViewRequest")
+			net.WriteInt(chip:EntIndex(), 32) -- For lookup in the request table when the owner answers
+			net.WriteFloat(CurTime() + 60)            -- For making requests expire on time clientside
+			net.WriteEntity(initiator)                -- The player attempting to view the E2
+			net.WriteString(truncName)                -- Name of the E2 so the owner knows what they're agreeing to
+			net.WriteEntity(chip)             -- Chip entity for validation clientside
+		net.Send(chip.player)
+	end
+
+	util.AddNetworkString("WireExpression2_OpenEditor")
+	util.AddNetworkString("WireExpression2_ViewRequest")
+	util.AddNetworkString("WireExpression2_AnswerRequest")
+	function TOOL:Think()
+		--[[
+			I had to replace TOOL:RightClick with TOOL:Think as prop protection was preventing
+			the view requests system from functioning as intended
+
+			So this manually handles right click meaning people don't need to give each other
+			full prop protection permissions in order to share a chip
+		]]
+		if not IsFirstTimePredicted() then return end
+
+		local player = self:GetOwner()
+		if player:KeyPressed(IN_ATTACK2) then
+			local chip = player:GetEyeTrace().Entity
+			if chip:IsPlayer() then return end
+
+			local player = self:GetOwner()
+
+			if IsValid(chip) and chip:GetClass() == "gmod_wire_expression2" then
+				if chip.player == player or bypassList[player:SteamID()] then -- Just download if the toolgun user owns this chip (or they're in the bypass list)
+					self:Download(player, chip)
```

Comment:
waitwaitwait, so basically what you're doing here is break prop protection in order to implement your thing?
big fuckin no!
We don't want wiremod to have PP bypass attack surface, ever.
Remove this, think of something else, work within the system.

Reply:
If you mean the `bypassList`, this was added due to people requesting a method for server operators to be able to take code without permissions to check for malicious chips, this is a server side concmd that takes vararg `STEAM_0` ids allowing server owners to give players capable of finding said malicious code a method to do so.

Reply:
not just the bypassList. I mean this entire function. This cannot be a Think hook since no PP will catch this. They will have to write custom code to deal with E2 and that'd mean even more attack surface.

Reply:
Well that's what view requests handles, there is no "attack surface" here, and I actually use a `CanTool` call as a fallback (forgot to do that until Divran pointed it out though).

Reply:
"there is no attack surface here" are some famous last words.

Reply:
@TomyLobo then please elaborate what the “attack surface” is here, given you cannot rely on a prop protection implementation being 100% secure in the first place.  

If you have any actual security concerns beyond sweeping statements, then I will address them immediately.

Reply:
bypassList also should not exist in wiremod, I would say. There is no support for "trusted" people actions (below Admin) anywhere else either, I think. 
It should probably be a hook CanInspectChip or the like (and maybe use IsAdmin() by default if the hook returns nil), which an admin mod can then implement however they see fit whether they have "trusted" moderators or just admins. Just my quick ideas.

Reply:
Well I can require anyone added by server operators (given this is a server only command that sets this list, only people with sv access can set it) to also be an admin, but given a server operator sets it I don’t see the point.  

As for allowing a wide range of people access to bypass the view request system (using role based permissions as opposed to the person by person permissions I added) doesn’t really fit with what people should actually be able to take code without permission, given the reason for someone taking code without permission would be to check for malicious code, then a player by player whitelist is required as you cannot expect every admin or above to have the skill required to do so, however you can expect quite a few (enough to warrant not giving ALL of them a bypass anyway) of them to steal code for the purpose of using it themselves or redistributing.

Reply:
Let me put it bluntly:
We want server owners to be able to allow their abusive admins to steal your code.
I think there's agreement about this among all the wire team members who weighed in on this so far and you still seem to try and negotiate about that.
What rights you want to give to whom is irrelevant here, Wiremod is just to provide the tools for the server owner to do so, in conjunction with a prop protection addon.

Reply:
> We want server owners to be able to allow their abusive admins to steal your code.
I think there's agreement about this among all the wire team members who weighed in on this so far and you still seem to try and negotiate about that.  

@TomyLobo are you sure you want to lump the entire wire team in with you on that?

Reply:
> We want server owners to be able to allow their abusive admins to steal your code.

nice

Reply:
> We want server owners to be able to allow their abusive admins to steal your code.

That's pretty blunt, but I mean yeah pretty much. 

At least personally, as a ~naughty~ vigilent admin, I'd be editing my copy of Wiremod to allow me to continue to skim through E2's if the consensus was to merge a version of this without built-in admin override. As this PR currently is, I guess I'd be whitelisting myself and all my admins, and then give it a shot - the prompt idea seems cool, might get a bit annoying if you have to keep approving the same friend though. I could see making a convar for enabling/disabling this system (for servers that want to keep the status-quo), though I'd imagine you'd prefer your system to be default-on to "protect against bad admins".

At one point I did indeed write something to save all uploaded E2's to the server, and debated uploading the whole local E2 folder (I think I was just too lazy to get it to work), initially as a way of ctrl-shift-f'ing for problematic code, but it came in handy when my server's backups proved more resilient than my mate's, and I was able to (shock him) by restoring his collection...